lib/srv/db/common/auth.go

```diff
@@ -1304,25 +1309,30 @@ func (r *awsRedisIAMTokenRequest) checkAndSetDefaults() error {
 // toSignedRequestURI creates a new AWS SigV4 pre-signed request URI.
 // This pre-signed request URI can then be used to authenticate as an
 // ElastiCache Redis or MemoryDB user.
-func (r *awsRedisIAMTokenRequest) toSignedRequestURI() (string, error) {
+func (r *awsRedisIAMTokenRequest) toSignedRequestURI(ctx context.Context) (string, error) {
 	if err := r.checkAndSetDefaults(); err != nil {
 		return "", trace.Wrap(err)
 	}
 	req, err := r.getSignableRequest()
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
-	s := v4.NewSigner(r.credentials)
-	_, err = s.Presign(req, nil, r.serviceName, r.region, time.Minute*15, r.clock.Now())
+	signer := v4.NewSigner()
+	creds, err := r.credProvider.Retrieve(ctx)
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
-	res := url.URL{
-		Host:     req.URL.Host,
-		Path:     "/",
-		RawQuery: req.URL.RawQuery,
+	signedURI, _, err := signer.PresignHTTP(ctx, creds, req, emptyPayloadHash, r.serviceName, r.region, r.clock.Now())
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if strings.HasPrefix(signedURI, "http://") {
+		return signedURI[len("http://"):], nil
+	}
+	if strings.HasPrefix(signedURI, "https://") {
+		return signedURI[len("https://"):], nil
 	}
-	return strings.TrimPrefix(res.String(), "//"), nil
+	return signedURI, nil
```

Comment:
nit: i think we knew it's http. if we do want to check HasPrefix, then we should return error if no schema match?

not sure if we can just use this lib?
https://github.com/build-on-aws/aws-redis-iam-auth-golang/blob/main/auth/auth.go
seems owned by amazon developers?

Reply:
lol, just noticed the `fmt.Println`. aws should really implement these signers properly in the sdk.